core-service/src/main/java/javasabr/mqtt/service/impl/InMemorySubscriptionService.java

```diff
@@ -60,23 +68,25 @@ public Array<SubscribeAckReasonCode> subscribe(
     return subscribeResults;
   }
 
-  private SubscribeAckReasonCode addSubscription(MqttUser user, MqttSession session, Subscription subscription) {
+  private SubscriptionResult addSubscription(MqttUser user, MqttSession session, Subscription subscription) {
     MqttClientConnectionConfig connectionConfig = user.connectionConfig();
     TopicFilter topicFilter = subscription.topicFilter();
     if (topicFilter.isInvalid()) {
-      return SubscribeAckReasonCode.TOPIC_FILTER_INVALID;
+      return INVALID_TOPIC_FILTER_RESULT;
     } else if (!connectionConfig.sharedSubscriptionAvailable() && topicFilter instanceof SharedTopicFilter) {
-      return SubscribeAckReasonCode.SHARED_SUBSCRIPTIONS_NOT_SUPPORTED;
+      return SHARED_SUBSCRIPTION_NOT_SUPPORTED_RESULT;
     } else if (!connectionConfig.wildcardSubscriptionAvailable() && topicFilter.wildcard()) {
-      return SubscribeAckReasonCode.WILDCARD_SUBSCRIPTIONS_NOT_SUPPORTED;
+      return WILDCARD_SUBSCRIPTION_NOT_SUPPORTED_RESULT;
     }
     ActiveSubscriptions activeSubscriptions = session.activeSubscriptions();
-    SingleSubscriber previous = subscriberTree.subscribe(user, subscription);
-    if (previous != null) {
-      activeSubscriptions.remove(previous.subscription());
+    SingleSubscriber newSubscriber = new SingleSubscriber(user, subscription);
+    SingleSubscriber previousSubscriber = subscriberTree.subscribe(newSubscriber);
+    boolean isSubscriptionAlreadyExisted = previousSubscriber != null;
```

Comment:
maybe we can return the reference to prev subscription?

Reply:
Such change makes the code dirtier, so let's keep as simple an abstraction as possible, unless, of course, you know of a specific case when a handler requires previous subscription.